From 7b5f53f548f62ad7d0df24c9e34fcc625950cbc6 Mon Sep 17 00:00:00 2001
From: "haitao.liu" <haitao.liu@amlogic.com>
Date: Wed, 2 Aug 2023 21:52:53 +0800
Subject: [PATCH 14/15] vdin: fix vdin memory access violation [1/1]

PD#SWPL-134178

Problem:
vdin set canvas width exceed active horizontal stride cause
memory access violation

Solution:
set canvas width equal to active horizontal stride

Verify:
t5m

Change-Id: I969a34a5e4dd2b6fe558285a507c7b9035c381fa
Signed-off-by: haitao.liu <haitao.liu@amlogic.com>
---
 drivers/media/vin/tvin/vdin/vdin_canvas.c | 8 ++++----
 drivers/media/vin/tvin/vdin/vdin_drv.h    | 2 +-
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/drivers/media/vin/tvin/vdin/vdin_canvas.c b/drivers/media/vin/tvin/vdin/vdin_canvas.c
index 7f81d0ee6..f72be7479 100644
--- a/drivers/media/vin/tvin/vdin/vdin_canvas.c
+++ b/drivers/media/vin/tvin/vdin/vdin_canvas.c
@@ -349,9 +349,6 @@ void vdin_canvas_auto_config(struct vdin_dev_s *devp)
 		break;
 	}
 
-	/*backup before roundup*/
-	devp->canvas_active_w = devp->canvas_w;
-
 	/* dw only support 8 bit mode */
 	if (devp->force_yuv444_malloc == 1 && !devp->double_wr) {
 		/* 4k is not support 10 bit mode in order to save memory */
@@ -362,7 +359,7 @@ void vdin_canvas_auto_config(struct vdin_dev_s *devp)
 			devp->canvas_w = h_active * VDIN_YUV444_8BIT_PER_PIXEL_BYTE;
 	}
 	/*canvas_w must ensure divided exact by 256bit(32byte)*/
-	devp->canvas_w = roundup(devp->canvas_w, devp->canvas_align);
+	devp->canvas_w = devp->canvas_active_w;
 	devp->canvas_h = v_active;
 
 	switch (devp->format_convert) {
@@ -584,8 +581,11 @@ unsigned int vdin_cma_alloc(struct vdin_dev_s *devp)
 				devp->vf_mem_size / devp->v_shrink_times;
 		else
 			devp->vf_mem_size_small = 0;
+		/* dw only support 8bit mode */
+		devp->canvas_active_w = devp->h_shrink_out * VDIN_YUV422_8BIT_PER_PIXEL_BYTE;
 	} else {
 		devp->vf_mem_size_small = 0;
+		devp->canvas_active_w = h_size;
 	}
 
 	if (devp->vf_mem_size_small)
diff --git a/drivers/media/vin/tvin/vdin/vdin_drv.h b/drivers/media/vin/tvin/vdin/vdin_drv.h
index 656f1e9a2..dddf873e8 100644
--- a/drivers/media/vin/tvin/vdin/vdin_drv.h
+++ b/drivers/media/vin/tvin/vdin/vdin_drv.h
@@ -821,7 +821,7 @@ struct vdin_dev_s {
 	unsigned int v_active_org;/*vdin scaler in*/
 	unsigned int canvas_h;
 	unsigned int canvas_w;
-	unsigned int canvas_active_w;
+	unsigned int canvas_active_w;/* mif stride */
 	unsigned int canvas_align_w;
 	unsigned int canvas_max_size;
 	unsigned int canvas_max_num;
-- 
2.25.1

